### PR TITLE
[Aptos Framework][vm] Aggregator V1

### PR DESCRIPTION
**Pull request by georgemitenkov**
### Description

This adds conversions from delta ops to write ops. Moreover, addition/subtraction return `PartialVMResult` that allows us to have uniform aborts both in Move code and in execution.

### Test Plan

---
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2462)

### CODE REVIEW

File: aptos-move/aptos-vm/src/delta_ext/delta_change_set.rs [hunk 1]
```diff
@@ -17,34 +30,80 @@ pub enum DeltaOp {
 }
 
 impl DeltaOp {
-    /// Returns optional result of delta application to `base` (None if
-    /// postocndition not satisfied).
-    pub fn apply_to(&self, base: u128) -> Option<u128> {
+    /// Returns the result of delta application to `base` or error if
+    /// postocndition is not satisfied.
```
Comment (gelash) [inline aptos-move/aptos-vm/src/delta_ext/delta_change_set.rs:34]: postcondition

**Review by zekun000 — APPROVED** [review]

**Comment by georgemitenkov** [comment]
@zekun000 @davidiw @movekevin @gelash so aggregator moved here (thanks to @wrwg 🎉 😆 ), any final thoughts before we land?

In short: currently we have aggregator as cfg feature, so all code should abort unless it's enabled. There is also a todo wrt gas cost, but this can be addressed later

File: aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs [hunk 2]
```diff
@@ -0,0 +1,722 @@
...
+    PositiveDelta,
+}
+
+/// Uniquely identifies each aggregator instance in storage.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct AggregatorID {
+    // A handle that is shared accross all aggregator instances created by the
+    // same `AggregatorFactory` and which is used for fine-grained storage
+    // access.
+    pub handle: u128,
```
Comment (zekun000) [inline aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs:46]: just fyi @msmouse is looking to change the handle to u256

File: aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs [hunk 3]
```diff
@@ -0,0 +1,722 @@
...
+    }
+}
+
+/// Represents a single aggregator change.
+#[derive(Debug)]
+pub enum AggregatorChange {
+    // A value should be written to storage.
+    Write(u128),
+    // A delta should be applied to storage.
+    Apply(DeltaOp),
```
Comment (zekun000) [inline aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs:213]: I like the name Merge more here

File: aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs [hunk 4]
```diff
@@ -0,0 +1,722 @@
...
+    /// Returns all changes made within this context (i.e. by a single
+    /// transaction).
+    pub fn into_change_set(self) -> AggregatorChangeSet {
+        let NativeAggregatorContext {
+            aggregator_data, ..
+        } = self;
+        let AggregatorData {
+            destroyed_aggregators,
+            aggregators,
+            ..
```
Comment (zekun000) [inline aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs:255]: I guess new_aggregators is only used to filter out created but destroyed ones in the same session

File: aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs [hunk 5]
```diff
@@ -0,0 +1,722 @@
...
+    // TODO: change this when we implement commutative subtraction.
+    aggregator.materialize(aggregator_context, &id)?;
+    aggregator.sub(value)?;
+
+    // TODO: charge gas properly.
+    Ok(NativeResult::ok(0, smallvec![]))
+}
+
+/// Move signature:
+/// fun remove_aggregator(table_handle: u128, key: u128);
```
Comment (zekun000) [inline aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs:463]: why this doesn't take an Aggregator instead?

**Review by zekun000 — APPROVED** [review]
looks great

File: aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs (hunk 5 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs:463]: No specific reason - changed that:)

File: aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs (hunk 4 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs:255]: Yes!

File: aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs (hunk 2 above)
Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs:46]: Once that lands I will make another PR to fix that on aggregator side:)

File: aptos-move/framework/aptos-experimental/sources/aggregator_factory.move [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+/// `aggregator.move`.
+///
+/// Recall that the value of any aggregator can be identified in storage by
+/// (handle, key) pair. How this pair can be generated? Short answer: with
+/// `AggregatorFactory`!
+///
+/// `AggregatorFactory` is a struct that can be stored as a resource on some
+/// account and which contains a `phantom_table` field. When the factory is
+/// initialized, we initialize this table. Importantly, table initialization
+/// only generates a uniue table `handle` - something we can reuse.
```
Comment (gelash) [inline aptos-move/framework/aptos-experimental/sources/aggregator_factory.move:15]: unique

File: aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs [hunk 7]
```diff
@@ -0,0 +1,736 @@
...
+    // Extract fields: `limit` of the new aggregator and a `phantom_handle` of
+    // the parent factory.
+    let limit = pop_arg!(args, u128);
+    let handle = get_handle(&pop_arg!(args, StructRef))?;
+
+    // Get the current aggregator data.
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+
+    // Every aggregator instance uses a unique key in its id. Here we can reuse
```
Comment (gelash) [inline aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs:334]: let's separate and test key generation logic (against separately computed what we should expect).
Additionally, let's make sure to confirm with any updates that happened to table handles (u256 discussion)

Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/move_vm_ext/aggregator_extension.rs:334]: I ll make a separate PR for that

**Pull request merged**